docs/design/design_v1.8.md

```diff
@@ -0,0 +1,437 @@
+
+#### Shared token discovery
+
+If `kubeadm join` is invoked with `--discovery-token`, token discovery is used; in this case the node basically retrieves the cluster CA certificates from the  `cluster-info` ConfigMap in the `kube-public` namespace.
+
+In order to prevent "man in the middle" attacks, several steps are taken:
+
+- First, the CA certificate is retrived via insecure connection (NB. this is possible because `kebeadm init` granted access to  `cluster-info` users for `system:unauthenticated` )
+- Then the CA certificate goes trough following validation steps: 
+  - "Basic validation", using the token ID against a JWT signature
+  - "Pub key validation", using provided `--discovery-token-ca-cert-hash`. This value is available in the output of "kubeadm init" or can be calcuated using standard tools (the hash is calculated over the bytes of the Subject Public Key Info (SPKI) object as in RFC7469). The `--discovery-token-ca-cert-hash flag`may be repeated multiple times to allow more than one public key.
+  - as a additional validation, the CA certificate is retrived via secure connection and then compared with the CA retrieved initially
+
+Please note that:
+
+1.  "Pub key validation" can be skipped passing `--discovery-token-unsafe-skip-ca-verification` flag; This weakens the kubeadm security model since others can potentially impersonate the Kubernetes Master.
+
+#### File/https discovery
+
+If `kubeadm join` is invoked with `--discovery-file`, file discovery is used; this file can be a local file or downloaded via an HTTPS URL; in case of HTTPS, the host installed CA bundle is used to verify the connection.
+
+With file discovery, the cluster CA certificates is provided into the file itself; however, as soon as the connection with the cluster is established, kubeadm try to access the `cluster-info` ConfigMap, and if available, uses it.
+
+## TLS Bootstrap
+
+Once the cluster info are known, TLS Boostrap is executed.
+
+The TLS bootstrap mechanism uses the shared token to temporarily authenticate with the Kubernetes Master to submit a certificate signing request (CSR) for a locally created key pair. 
+
+The request is then automatically approved and the operation completes saving `ca.crt` file and `kubelet.conf` file to be used by kubelet for joining the cluster.
+
+Please note that:
+
+- The temporary authentication is validated against the token saved during the `kubeadm init` process (or with additional tokens created with `kuubeadm token`) 
+- The temporary authentication resolve to a user member of  `system:bootstrappers:kubeadm:default-node-token` group (formerly `system:bootstrappers` in v1.7) which was granted access to CSR api during the `kubeadm init` process
+- The automatic CSR approval is managed by the csrapprover controller, according with configuration done the `kubeadm init` process
+
+**TODO:** It is planned that the kubelet will handle this phase 2 on its own in v1.9
```

Comment:
This is already the case in v1.8
kubeadm doesn't actually do the TLS bootstrap anymore for v1.8 clusters, that's delegated to the kubelet